ffi/derive/src/convert.rs

```diff
@@ -525,29 +762,130 @@ fn derive_into_ffi_for_fieldless_enum(enum_name: &Ident, repr: &[syn::NestedMeta
     }
 }
 
+fn derive_transparent(
+    name: &Ident,
+    generics: &syn::Generics,
+    item: &syn::DataStruct,
+) -> TokenStream2 {
+    let inner = item
+        .fields
+        .iter()
+        .next()
+        .map(|field| &field.ty)
+        .expect_or_abort("transparent struct must have at least one field");
+
+    let lifetime: syn::Lifetime = syn::parse_quote!('__iroha_ffi_itm);
+    let mut generics = generics.clone();
+    let (generic_params, ty_generics, where_clause) =
+        add_bounds_and_split_generics(&mut generics, &[], &[], &[]);
+
+    quote! {
+        // Note: if `#inner` is generic or doesn't have the same layout with `#name` this won't compile
+        impl <#generic_params> iroha_ffi::Transparent for #name #ty_generics #where_clause {
+            type Inner = #inner;
+
+            fn into_inner(outer: Self) -> Self::Inner {
+                unsafe { core::mem::transmute(outer) }
+            }
+
+            fn from_inner(inner: Self::Inner) -> Self {
+                unsafe { core::mem::transmute(inner) }
+            }
+        }
+
+        impl <#lifetime, #generic_params> iroha_ffi::Transparent for &#lifetime #name #ty_generics #where_clause {
+            type Inner = &#lifetime #inner;
+
+            fn into_inner(outer: Self) -> Self::Inner {
+                unsafe { & *(outer as *const _ as *const _) }
+            }
+
+            fn from_inner(inner: Self::Inner) -> Self {
+                unsafe { & *(inner as *const _ as *const _) }
+            }
+        }
```

Comment:
I don't like approach with `Transparent` traits because it gives user the option to implement different `Inner` type then the type that actually inside the transparent struct. Especially, when all of the information to derive correct behavior is given in the type definition and you could just delegate to the inner type.

Could you derive `IntoFfi` and `TryFromReprC` directly without the `Transparent` trait? Something like:
```rs
        impl<#generic_params> iroha_ffi::IntoFfi for #name #ty_generics #where_clause {
            type Target = <#inner_type as iroha_ffi::IntoFfi>::Target;

            fn into_ffi(self) -> Self::Target {
                iroha_ffi::IntoFfi::into_ffi(self.0)
            }
```

Also, why do you need to special case implementations for references as in `&#name`, etc?

Reply:
I will try to explain my intentions here.
Maybe you noticed kinda strange signature that we have:
```rust
impl<T: IntoFfi> IntoFfi for Type where Self: Transparent<Inner = T> {
     ...
}
```
Problem is following: suppose we have some kind of newtype `Newtype(Inner)`. By just looking at `Inner` we can't know if `Inner`, `&Inner`, `&mut Inner` implement `IntoFfi`.
So by introducing generic parameter we can move error invocation from `impl` block to trying actually call unsupported conversion.

For example currently we don't implement `IntoFfi` for `&String/&mut String`, so:
```rust
// Without generic parameter expansion would always fail 
// because we can't know that we need to derive only `Name: IntoFfi` but not `&Name/&mut Name: IntoFfi`.
// However with generic param everything would generate without problem and ... 
#[derive(IntoFfi, TryFromReprC)]
#[repr(transparent)]
struct Name(String);

fn main() {
    /// only trying to call unimplemented method would fail 
    /// because there no such type `T` that both `&Name: Transparent<Inner = &T>` and `&T: IntoFfi`.
    let name_ref_ffi = IntoFfi::into_ffi(&name);
}
```

I think without `Tranparent` trait we can't actually express this i believe. 
I agree with you that it is not great to expose kinda inner trait to public interface.

One small notice: `transmute` are kind of guard here :) because it's checks that size of `Name` and `String` are the same. However i understand that `transmute` is really unsafe and try to use other conversions where possible.

Reply:
>  I agree with you that it is not great to expose inner trait to public interface.

my main concern is that it's getting complicated and offering forms of extensibility that will never get to be used.

> Problem is following: suppose we have some kind of newtype `Newtype(Inner)`. By just looking at Inner we can't know if `Inner`, `&Inner`, `&mut Inner` implement `IntoFfi`.
For example currently we don't implement IntoFfi for &String/&mut String

Ok, I get you. While we could require that `&T: IntoFfi`, in many cases it's not possible to give a sound `&mut T` implementation (`String` is a good example)

Reply:
> I think without Tranparent trait we can't actually express this i believe.

After some thought, I have to agree

> What we can do is to mark this trait and every it's method as unsafe to warn user to be extra careful.

unsafe on traits is a marker for those implementing the trait that they have to obey certain invariants when implementing the trait while unsafe on methods is a marker for those calling the method that they have to uphold certain invariants when calling the method. From the looks of it, I think that trait should be marked unsafe, but not the methods

Reply:
> From the looks of it, I think that trait should be marked unsafe, but not the methods

Yes, i agree with you on that.

Reply:
> After some thought, I have to agree. The only alternative would be to offload the decision to the user via derive macro helper attributes

I was thinking about helper attributes but in different way: make `derive(Transparent)` macro with helper attribute on the inner field. 
Also this macro could check that struct indeed have `repr(Transparent)` attribute and place kind some kind of helper attribute like `#[transparent]` on it to check it from `IntoFfi/TryFromFfi`.

This would allow user to decide how them what to pass transparent struct by unwrapping inner field or by opaque pointer e.g. despite `Name` is transparent right now we cant derive `IntoFfi/TryFromFfi` on it because `ConstString` is union and unions is not supported right now. 

However this whole thing seems to be overkill.

Reply:
This sounds like an `X` / `Y` problem. All we need is to be able to use transparent structures. A manual `impl` for `Name` at this time is very much not an issue, and given how it's implemented, definitely not an issue of code size. I recommend postponing this disucssion after dynamic linking is at least somewhat functional. We've spent a lot of time doing the FFI, we need to show something for it.

Reply:
Ok, i will resolve this discussion than, i have created #2669.